### PR TITLE
Upgrade Chalk

### PR DESCRIPTION
**Pull request by jackh726**
~~Blocked on rust-lang/chalk#670~~
~~Now blocked on rust-lang/chalk#680 and release~~

In addition to the straight upgrade, I also tried to fix some tests by properly returning variables and max universes in the solution. Unfortunately, this actually triggers the same perf problem that rustc traits code runs into in `canonicalizer`. Not sure what the root cause of this problem is, or why it's supposed to be solved in chalk.

r? @nikomatsakis

### CODE REVIEW

File: src/tools/tidy/src/extdeps.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@ pub fn check(root: &Path, bad: &mut bool) {
 
         // Ensure source is allowed.
         if !ALLOWED_SOURCES.contains(&&*source) {
-            tidy_error!(bad, "invalid source: {}", source);
+            //tidy_error!(bad, "invalid source: {}", source);
```
Comment (jackh726) [inline src/tools/tidy/src/extdeps.rs:30]: This was only needed so I can test locally with the git deps; I'll remove it when we have a real usable chalk version upstream.

**Comment by jackh726** [comment]
I'll fix the merge conflict when this is no longer blocked.

**Comment by JohnCSimon** [comment]
@Rustbot label: -S-waiting-on-review +S-waiting-on-author

**Comment by jackh726** [comment]
@rustbot label: -S-waiting-on-author +S-blocked

This is waiting on Chalk.

**Comment by jackh726** [comment]
welp, messed up submodules in rebase; will fix when chalk blockers are fixed

File: src/tools/tidy/src/extdeps.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/tools/tidy/src/extdeps.rs:30]: did you to remove this before it lands?

**Review by nikomatsakis — APPROVED** [review]
r=me modulo the comment by @jackh726

**Comment by nikomatsakis** [comment]
@jackh726 feel free to r=nikomatsakis at some point

File: src/tools/tidy/src/extdeps.rs (hunk 1 above)
Comment (jackh726) [reply src/tools/tidy/src/extdeps.rs:30]: Ugh whoops. Forgot about this. Will do that in a bit.`

**Pull request merged**